File: support-models/src/main/scala/com/gu/support/workers/CheckoutFailureReasons.scala
package com.gu.support.workers

import io.circe.{Decoder, Encoder}

import scala.PartialFunction.condOpt

object CheckoutFailureReasons {

  val all = List(
    InsufficientFunds,
    PaymentMethodDetailsIncorrect,
    PaymentMethodTemporarilyDeclined,
    PaymentMethodUnacceptable,
    PaymentProviderUnavailable,
    PaymentRecentlyTaken,
    AccountMismatch,
    AmazonPayTryAnotherCard,
    AmazonPayTryAgain,
    AmazonPayFatal,
    StripePaymentMethodDisabled,
    Unknown,
  )

  def fromString(string: String): Option[CheckoutFailureReason] = all.find(_.asString == string)

  sealed trait CheckoutFailureReason {
    def asString: String
  }

  case object StripePaymentMethodDisabled extends CheckoutFailureReason {
    override def asString: String = "Stripe payments are currently disabled"
  }

  case object InsufficientFunds extends CheckoutFailureReason {
    override def asString: String = "insufficient_funds"
  }

  case object PaymentMethodDetailsIncorrect extends CheckoutFailureReason {
    override def asString: String = "payment_details_incorrect"
  }

  case object PaymentMethodTemporarilyDeclined extends CheckoutFailureReason {
    override def asString: String = "payment_method_temporarily_declined"
  }

  case object PaymentMethodUnacceptable extends CheckoutFailureReason {
    override def asString: String = "payment_method_unacceptable"
  }

  case object PaymentProviderUnavailable extends CheckoutFailureReason {
    override def asString: String = "payment_provider_unavailable"
  }

  case object PaymentRecentlyTaken extends CheckoutFailureReason {
    override def asString: String = "payment_recently_taken"
  }

  case object AccountMismatch extends CheckoutFailureReason {
    override def asString = "production_test_account_mismatch"
  }

  case object AmazonPayTryAnotherCard extends CheckoutFailureReason {
    override def asString: String = "amazon_pay_try_other_card"
  }

  case object AmazonPayTryAgain extends CheckoutFailureReason {
    override def asString: String = "amazon_pay_try_again"
  }

  case object AmazonPayFatal extends CheckoutFailureReason {
    override def asString: String = "amazon_pay_fatal"
  }

  case object Unknown extends CheckoutFailureReason {
    override def asString: String = "unknown"
  }

  // https://stripe.com/docs/declines/codes
  def convertStripeDeclineCode(declineCode: String): Option[CheckoutFailureReason] = condOpt(declineCode) {
    case "approve_with_id" => PaymentMethodTemporarilyDeclined
    case "call_issuer" => PaymentMethodUnacceptable
    case "card_not_supported" => PaymentMethodUnacceptable
    case "card_velocity_exceeded" => PaymentMethodUnacceptable
    case "currency_not_supported" => PaymentMethodUnacceptable
    case "do_not_honor" => PaymentMethodUnacceptable
    case "do_not_try_again" => PaymentMethodUnacceptable
    case "duplicate_transaction" => PaymentRecentlyTaken
    case "expired_card" => PaymentMethodUnacceptable
    case "fraudulent" => PaymentMethodUnacceptable
    case "generic_decline" => PaymentMethodUnacceptable
    case "incorrect_number" => PaymentMethodDetailsIncorrect
    case "incorrect_cvc" => PaymentMethodDetailsIncorrect
    case "incorrect_pin" => PaymentMethodDetailsIncorrect
    case "incorrect_zip" => PaymentMethodDetailsIncorrect
    case "insufficient_funds" => InsufficientFunds
    case "invalid_account" => PaymentMethodUnacceptable
    case "invalid_amount" => PaymentMethodUnacceptable
    case "invalid_cvc" => PaymentMethodDetailsIncorrect
    case "invalid_expiry_year" => PaymentMethodDetailsIncorrect
    case "invalid_number" => PaymentMethodDetailsIncorrect
    case "invalid_pin" => PaymentMethodDetailsIncorrect
    case "issuer_not_available" => PaymentMethodTemporarilyDeclined
    case "lost_card" => PaymentMethodUnacceptable
    case "new_account_information_available" => PaymentMethodUnacceptable
    case "no_action_taken" => PaymentMethodUnacceptable
    case "not_permitted" => PaymentMethodUnacceptable
    case "pickup_card" => PaymentMethodUnacceptable
    case "pin_try_exceeded" => PaymentMethodUnacceptable
    case "processing_error" => PaymentMethodTemporarilyDeclined
    case "reenter_transaction" => PaymentMethodTemporarilyDeclined
    case "restricted_card" => PaymentMethodUnacceptable
    case "revocation_of_all_authorizations" => PaymentMethodUnacceptable
    case "revocation_of_authorization" => PaymentMethodUnacceptable
    case "security_violation" => PaymentMethodUnacceptable
    case "service_not_allowed" => PaymentMethodUnacceptable
    case "stolen_card" => PaymentMethodUnacceptable
    case "stop_payment_order" => PaymentMethodUnacceptable
    case "transaction_not_allowed" => PaymentMethodUnacceptable
    case "try_again_later" => PaymentMethodTemporarilyDeclined
    case "withdrawal_count_limit_exceeded" => PaymentMethodUnacceptable
    case "Stripe payments are currently disabled" => StripePaymentMethodDisabled
  }

  def convertAmazonPayDeclineCode(declineCode: String): CheckoutFailureReason = declineCode match {
    case "InvalidPaymentMethod" => AmazonPayTryAnotherCard
    case "ProcessingFailure" => AmazonPayTryAgain
    case _ => AmazonPayFatal
  }

  implicit val encodeFailureReason: Encoder[CheckoutFailureReason] =
    Encoder.encodeString.contramap[CheckoutFailureReason](_.asString)

  implicit val decodeFailureReason: Decoder[CheckoutFailureReason] = Decoder.decodeString.emap { identifier =>
    CheckoutFailureReasons.fromString(identifier).toRight(s"Unrecognised failure reason '$identifier'")
  }

}

File: support-workers/src/typescript/errors/checkoutFailureReasons.ts
export const checkoutFailureReasons = [
	'insufficient_funds',
	'payment_details_incorrect',
	'payment_method_temporarily_declined',
	'payment_method_unacceptable',
	'payment_recently_taken',
	'unknown',
] as const;
export type CheckoutFailureReason = (typeof checkoutFailureReasons)[number];

export const stripeDeclineCodeMap: Record<string, CheckoutFailureReason> = {
	approve_with_id: 'payment_method_temporarily_declined',
	call_issuer: 'payment_method_unacceptable',
	card_not_supported: 'payment_method_unacceptable',
	card_velocity_exceeded: 'payment_method_unacceptable',
	currency_not_supported: 'payment_method_unacceptable',
	do_not_honor: 'payment_method_unacceptable',
	do_not_try_again: 'payment_method_unacceptable',
	duplicate_transaction: 'payment_recently_taken',
	expired_card: 'payment_method_unacceptable',
	fraudulent: 'payment_method_unacceptable',
	generic_decline: 'payment_method_unacceptable',
	incorrect_number: 'payment_details_incorrect',
	incorrect_cvc: 'payment_details_incorrect',
	incorrect_pin: 'payment_details_incorrect',
	incorrect_zip: 'payment_details_incorrect',
	insufficient_funds: 'insufficient_funds',
	invalid_account: 'payment_method_unacceptable',
	invalid_amount: 'payment_method_unacceptable',
	invalid_cvc: 'payment_details_incorrect',
	invalid_expiry_year: 'payment_details_incorrect',
	invalid_number: 'payment_details_incorrect',
	invalid_pin: 'payment_details_incorrect',
	issuer_not_available: 'payment_method_temporarily_declined',
	lost_card: 'payment_method_unacceptable',
	new_account_information_available: 'payment_method_unacceptable',
	no_action_taken: 'payment_method_unacceptable',
	not_permitted: 'payment_method_unacceptable',
	pickup_card: 'payment_method_unacceptable',
	pin_try_exceeded: 'payment_method_unacceptable',
	processing_error: 'payment_method_temporarily_declined',
	reenter_transaction: 'payment_method_temporarily_declined',
	restricted_card: 'payment_method_unacceptable',
	revocation_of_all_authorizations: 'payment_method_unacceptable',
	revocation_of_authorization: 'payment_method_unacceptable',
	security_violation: 'payment_method_unacceptable',
	service_not_allowed: 'payment_method_unacceptable',
	stolen_card: 'payment_method_unacceptable',
	stop_payment_order: 'payment_method_unacceptable',
	transaction_not_allowed: 'payment_method_unacceptable',
	try_again_later: 'payment_method_temporarily_declined',
	withdrawal_count_limit_exceeded: 'payment_method_unacceptable',
};

export function checkoutFailureReasonFromErrorMessage(
	errorMessage: string,
): CheckoutFailureReason {
	for (const [declineCode, reason] of Object.entries(stripeDeclineCodeMap)) {
		if (errorMessage.includes(declineCode)) {
			return reason;
		}
	}
	return 'unknown';
}
